### PR TITLE
Bug 5571/v12

### PR DESCRIPTION
**Pull request by victorjulien**
Action handling overhaul. Needs review and more tests. PR for review and seeing which parts codecov shows as untested.

Includes minor cleanups.

suricata-verify-pr: 962

replaces #8013

### CODE REVIEW

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW1_stats_chk | | |
| .tcp.rst | 128100 | 104005 | 81.19% |
| | SURI_TLPR1_stats_chk | | |
| .flow.spare | 2202676 | 2037740 | 92.51% |
| .flow.memuse | 557245568 | 607502848 | 109.02% |
| | TREX_GENERIC_stats_chk | | |
| .capture.kernel_drops | 0 | 488875 | - |
| .flow.end.state.new | 6272 | 9179 | 146.35% |
| .flow.end.tcp_state.syn_sent | 0 | 804 | - |
| .flow.end.tcp_state.syn_recv | 0 | 12 | - |
| .flow.end.tcp_state.fin_wait1 | 0 | 124 | - |
| .flow.end.tcp_state.fin_wait2 | 0 | 46 | - |
| .flow.end.tcp_state.time_wait | 0 | 71 | - |
| .flow.end.tcp_state.last_ack | 0 | 32 | - |
| .flow.end.tcp_state.close_wait | 0 | 183 | - |
| .flow.end.tcp_liberal | 80952 | 102837 | 127.03% |
| .tcp.reassembly_gap | 80952 | 147272 | 181.93% |
| .tcp.overlap | 0 | 8 | - |
| .app_layer.error.http.parser | 0 | 39 | - |
| .app_layer.error.ftp.gap | 0 | 7 | - |
| .app_layer.error.smtp.gap | 0 | 114 | - |
| .app_layer.error.tls.gap | 53852 | 73274 | 136.07% |
| .app_layer.error.dcerpc_tcp.parser | 0 | 35 | - |
| .ftp.memuse | 0 | 200 | - | 

Pipeline 10019

**Comment by jufajardini** [comment]
Tried to understand all changes, will need to re-check a few parts to properly grasp the changes, still. 

I think this will have to be merged first, or at least approved, before working on 5180 (alert queue optimizations/not missing drop action even if there's a malloc failure) can continue...

File: src/decode.h [hunk 1]
```diff
@@ -778,58 +768,33 @@ void CaptureStatsSetup(ThreadVars *tv, CaptureStats *s);
...
+static inline void PacketDrop(Packet *p, const uint8_t action, enum PacketDropReason r)
 {
-    if (likely(p->root == NULL)) {
-        PACKET_SET_ACTION(p, a);
-    } else {
-        PACKET_SET_ACTION(p->root, a);
+    if (p->drop_reason == PKT_DROP_REASON_NOT_SET)
+        p->drop_reason = (uint8_t)r;
+    if (p->root) {
+        p->root->action |= (action & ACTION_DROP_REJECT);
```
Comment (victorjulien) [inline src/decode.h:776]: remove mask, we never set action flags other than drop/reject anymore (debug validation can check that)

**Comment by ct0br0** [comment]
capture kernel_drops

For reference

**Comment by victorjulien** [comment]
replaced by #8030

**Pull request closed without merging**